server/src/main/java/org/candlepin/policy/activationkey/ActivationKeyRules.java

```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright (c) 2009 - 2012 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.policy.activationkey;
+
+import org.candlepin.model.ConsumerType;
+import org.candlepin.model.Pool;
+import org.candlepin.model.activationkeys.ActivationKey;
+import org.candlepin.model.activationkeys.ActivationKeyPool;
+import org.candlepin.policy.ValidationResult;
+
+import com.google.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xnap.commons.i18n.I18n;
+
+import java.util.Optional;
+
+/**
+ * ActivationKeyRules
+ *
+ * Enforces rules similar to EntitlementRules.java for
+ * attaching pools to activation keys
+ */
+public class ActivationKeyRules {
+
+    private static Logger log = LoggerFactory.getLogger(ActivationKeyRules.class);
+
+    private I18n i18n;
+
+    @Inject
+    public ActivationKeyRules(I18n i18n) {
+        this.i18n = i18n;
+    }
+
+    /**
+     * Checks that the specified pool can be attached to the specified activation key. A result object
+     * is returned which contains errors/warnings, or is empty if validation did not find any issues.
+     *
+     * @param key The activation key to attach the pool to
+     *
+     * @param pool The pool to be attached to the activation key
+     *
+     * @param quantity The quantity of the pool the key will consume
+     *
+     * @return a result object which contains errors or warnings produced by the validation process.
+     *  In case the validation was successful, it will be empty.
+     */
+    public ValidationResult runPoolValidationForActivationKey(ActivationKey key, Pool pool, Long quantity) {
+        ValidationResult result = new ValidationResult();
+        if (quantity != null && quantity < 1) {
+            result.addError("rulefailed.invalid.quantity");
```

Comment:
Now that we're not passing the errors through the Java->JS layer, we should probably transition these to a proper enum. If we still need these error strings, we can include them as a property of the enum. As an added bonus, we'll get a tiny performance uplift in all the string checking when we build proper messages, as we could also attach *those* to the enums as like a "buildErrorMessage" method:
```
public enum ValidationErrors { // rename as appropriate
  ALREADY_EXISTS("rulefailed.already.exists", "Multi-entitlement not supported for pool \"{0}\"");
  // Add other errors here

  private final String code;
  private final String errmsg;

  ValidationError(String code, String errmsg) {
    this.code = code;
    this.errmsg = errmsg;
  }

  public String toString() {
    return this.code;
  }

  public String buildErrorMessage(I18n i18n, Object... args) {
    return i18n.tr(this.errmsg, args);
  }
}
```

Reply:
Done. ValidationResult now will also accept Enums as error keys, and eventually, as more rules move to Java, we'll get rid of ValidationError/ValidationWarning classes altogether.